internal/controller/bucket_controller.go

```diff
@@ -0,0 +1,225 @@
+	managementcluster.ManagementCluster
+}
+
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the bucket closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
+func (r BucketReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Started reconciling Bucket")
+	defer logger.Info("Finished reconciling Bucket")
+
+	// Get the bucket that we are reconciling
+	bucket := &v1alpha1.Bucket{}
+	err := r.Client.Get(ctx, req.NamespacedName, bucket)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
+	}
+
+	// Create the correct service implementation based on the provider
+	var service objectstorage.ObjectStorageService
+	switch r.ManagementCluster.Provider {
+	case "capa":
+		roleArn, err := r.getRoleArn(ctx)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+
+		service, err = r.ObjectStorageServiceFactory.NewS3Service(ctx, r.ManagementCluster.Region, roleArn)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	default:
+		return ctrl.Result{}, errors.New(fmt.Sprintf("unsupported provider %s", r.ManagementCluster.Provider))
+	}
+
+	// Handle deleted clusters
+	if !bucket.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, service, bucket)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, service, bucket)
+}
+
+// reconcileCreate creates the s3 bucket.
+func (r BucketReconciler) reconcileNormal(ctx context.Context, service objectstorage.ObjectStorageService, bucket *v1alpha1.Bucket) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	originalBucket := bucket.DeepCopy()
+	// If the Bucket doesn't have our finalizer, add it.
+	if controllerutil.AddFinalizer(bucket, v1alpha1.BucketFinalizer) {
+		// Register the finalizer immediately to avoid orphaning AWS resources on delete
+		if err := r.Client.Patch(ctx, bucket, client.MergeFrom(originalBucket)); err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	logger.Info("Checking if bucket exists")
+	exists, err := service.ExistsBucket(ctx, bucket)
+	if err != nil {
+		logger.Info(fmt.Sprintf("Either you don't have access to bucket %v or another error occurred. "+
+			"Here's what happened: %v", bucket.Spec.Name, err))
+		return ctrl.Result{}, errors.WithStack(err)
+	} else if !exists {
+		logger.Info(fmt.Sprintf("Bucket %v is available, creating", bucket.Spec.Name))
```

Comment:
You can add the `bucket.Spec.Name` to the logger with this to avoid putting it constantly in the message
```go
logger = logger.WithValues("bucket-name", bucket.Spec.Name)
```

Reply:
Nice thank you